discovery_engine_core/core/src/state.rs

```diff
@@ -108,3 +110,163 @@ impl Engine {
+        pub(super) topic: String,
+    }
+
+    impl From<NewsResourceWithNaiveDateTime> for NewsResource {
+        fn from(resource: NewsResourceWithNaiveDateTime) -> Self {
+            NewsResource {
+                title: resource.title,
+                snippet: resource.snippet,
+                url: resource.url,
+                source_domain: resource.source_domain,
+                date_published: DateTime::<Utc>::from_utc(resource.date_published, Utc),
+                image: resource.image,
+                rank: resource.rank,
+                score: resource.score,
+                country: resource.country,
+                language: resource.language,
+                topic: resource.topic,
+            }
+        }
+    }
+
+    pub(crate) fn deserialize(bytes: &[u8]) -> Result<HashMap<StackId, Data>, bincode::Error> {
+        bincode::deserialize::<HashMap<StackId, DataWithNaiveDateTime>>(bytes).map(|stacks| {
+            stacks
+                .into_iter()
+                .map(|(id, data)| (id, data.into()))
+                .collect()
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::collections::HashMap;
+
+    use chrono::{DateTime, NaiveDate, Utc};
+    use url::Url;
+    use xayn_discovery_engine_ai::Embedding;
+
+    use super::naive_date_time_migration::DataWithNaiveDateTime;
+    use crate::{
+        document::Id,
+        stack::{Data, Id as StackId},
+        state::naive_date_time_migration::{
+            self,
+            DocumentWithNaiveDateTime,
+            NewsResourceWithNaiveDateTime,
+        },
+    };
+
+    #[test]
+    fn test_naive_date_time_migration() {
+        let stack_id = StackId::nil();
+        let date_published = NaiveDate::from_ymd(2016, 7, 8).and_hms(10, 25, 55);
+        let old_document = DocumentWithNaiveDateTime {
+            id: Id::new(),
+            stack_id,
+            reaction: None,
+            smbert_embedding: Embedding::default(),
+            resource: NewsResourceWithNaiveDateTime {
+                title: String::default(),
+                snippet: String::default(),
+                url: Url::parse("https://example.net").unwrap(/* used only in tests */),
+                source_domain: "example.com".to_string(),
+                image: None,
+                score: None,
+                rank: 0,
+                country: "GB".to_string(),
+                language: "en".to_string(),
+                topic: "news".to_string(),
+                date_published,
+            },
+        };
+        let old_data = DataWithNaiveDateTime {
+            alpha: 1.,
+            beta: 1.,
+            likes: 1.,
+            dislikes: 1.,
+            documents: vec![old_document],
+        };
+        let old_stacks_data = HashMap::from([(stack_id, old_data)]);
+        let bytes = bincode::serialize(&old_stacks_data).unwrap();
+        let failing_op = bincode::deserialize::<HashMap<StackId, Data>>(&bytes);
+
+        assert!(failing_op.is_err());
```

Comment:
:art: Not sure if it's worth it to check that is failing

Reply:
I thought it might be good to prove it's failing without migration, should I remove it?